File: examples/lua/console/master.lua
-- Copyright (c) 2011, 2013-2014. The SimGrid Team.
-- All rights reserved.

-- This program is free software; you can redistribute it and/or modify it
-- under the terms of the license (GNU LGPL) which comes with this package.

--Master Function
function Master(...)

  if select("#", ...) ~= 4 then
    error("Wrong number of arguments (got " .. select("#", ...) ..
        ", expected 4: nb_tasks comp_size comm_size slave_count)")
  end

  simgrid.info("Hello from lua, I'm the master")
  for i,v in ipairs({...}) do
      simgrid.info("Got " .. v)
  end

  local nb_task, comp_size, comm_size, slave_count = select(1, ...)

  simgrid.info("Argc=" .. select("#", ...) .. " (should be 4)")

  -- Dispatch the tasks

  for i = 1, nb_task do
    task = simgrid.task.new("Task " .. i, comp_size, comm_size);
    local task_name = simgrid.task.get_name(task)
    alias = "slave " .. string.format("%d",i%slave_count);
    simgrid.info("Master sending  '" .. task_name .. "' To '" .. alias .. "'");
    simgrid.task.send(task, alias); -- C user data set to NULL
    simgrid.info("Master done sending '" .. task_name .. "' To '" .. alias .. "'");
  end

  -- Sending Finalize Message To Others

  simgrid.info("Master: All tasks have been dispatched. Let's tell everybody the computation is over.");
  for i = 0, slave_count-1 do
    alias = "slave " .. i;
    simgrid.info("Master: sending finalize to " .. alias);
    finalize = simgrid.task.new("finalize", comp_size, comm_size);
    simgrid.task.send(finalize, alias)
  end
  simgrid.info("Master: Everything's done.");
end
--end_of_master

File: examples/java/commTime/Master.java
/* Copyright (c) 2006-2014. The SimGrid Team.
 * All rights reserved.                                                     */

/* This program is free software; you can redistribute it and/or modify it
 * under the terms of the license (GNU LGPL) which comes with this package. */

package commTime;

import org.simgrid.msg.Host;
import org.simgrid.msg.Msg;
import org.simgrid.msg.MsgException;
import org.simgrid.msg.Process;
import org.simgrid.msg.Task;

public class Master extends Process {
  public Master(Host host, String name, String[]args) {
    super(host,name,args);
  }

  public void main(String[] args) throws MsgException {
    if (args.length < 4) {
      Msg.info("Master needs 4 arguments");
      System.exit(1);
    }

    int tasksCount = Integer.valueOf(args[0]).intValue();    
    double taskComputeSize = Double.valueOf(args[1]).doubleValue();    
    double taskCommunicateSize = Double.valueOf(args[2]).doubleValue();

    int slavesCount = Integer.valueOf(args[3]).intValue();

    Msg.info("Hello! Got "+  slavesCount + " slaves and "+tasksCount+" tasks to process");

    for (int i = 0; i < tasksCount; i++) {
      Task task = new Task("Task_" + i, taskComputeSize, taskCommunicateSize);
      if (i%1000==0)
         Msg.info("Sending \"" + task.getName()+ "\" to \"slave_" + i % slavesCount + "\"");
         task.send("slave_"+(i%slavesCount));
       }

    Msg.info("All tasks have been dispatched. Let's tell everybody the computation is over.");

    for (int i = 0; i < slavesCount; i++) {
      FinalizeTask task = new FinalizeTask();
      task.send("slave_"+(i%slavesCount));
    }
    Msg.info("Goodbye now!");
  }
}
